File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
*/

class Solution {
    int m;
    public int diameterOfBinaryTree(TreeNode root) {
        m = 0; // global maximum
        postOrderDFS(root); // return height of tree
        return m;
    }
    private int postOrderDFS(TreeNode root) {
        if (root == null) return 0;
        int left = postOrderDFS(root.left); // height of left subtree
        int right = postOrderDFS(root.right); // height of right subtree
        int tallestSubtreeHeight = Math.max(left, right);
        
        m = Math.max(left+right, m);
        // updating global maximum in case left+right is long enough
        // note the absence of the trunk of the tree, I'm just counting the 
        
        int height = tallestSubtreeHeight + 1; 
        // add current node to height
        // this is where the postorder magic happens
        // remove that + 1 and you'll get zero every time
        return height;
    }
       
}

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var maxLength: Int = 0;
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        maxLength = 0
        postOrderDFS(root) // height of tree
        return maxLength
    }
    private fun postOrderDFS(root: TreeNode?): Int {
        if (root == null) return 0
        val left = postOrderDFS(root.left) // height of left subtree
        val right = postOrderDFS(root.right) // height of right subtree
        
        maxLength = Math.max(left+right, maxLength)
        
        val tallestSubtree = Math.max(left, right)
        return tallestSubtree + 1
    }
}